server/src/infra/auth_service.rs

```diff
@@ -48,22 +48,27 @@ fn create_jwt(key: &Hmac<Sha512>, user: String, groups: HashSet<GroupIdAndName>)
     jwt::Token::new(header, claims).sign_with_key(key).unwrap()
 }
 
-fn get_refresh_token_from_cookie(
-    request: HttpRequest,
-) -> std::result::Result<(u64, String), HttpResponse> {
+fn parse_refresh_token(token: &str) -> std::result::Result<(u64, String), HttpResponse> {
+    match token.split_once('+') {
+        None => Err(HttpResponse::Unauthorized().body("Invalid refresh token")),
+        Some((token, u)) => {
+            let refresh_token_hash = {
+                let mut s = DefaultHasher::new();
+                token.hash(&mut s);
+                s.finish()
+            };
+            Ok((refresh_token_hash, u.to_string()))
+        }
+    }
+}
+
+fn get_refresh_token(request: HttpRequest) -> std::result::Result<(u64, String), HttpResponse> {
     match request.cookie("refresh_token") {
-        None => Err(HttpResponse::Unauthorized().body("Missing refresh token")),
-        Some(t) => match t.value().split_once('+') {
-            None => Err(HttpResponse::Unauthorized().body("Invalid refresh token")),
-            Some((token, u)) => {
-                let refresh_token_hash = {
-                    let mut s = DefaultHasher::new();
-                    token.hash(&mut s);
-                    s.finish()
-                };
-                Ok((refresh_token_hash, u.to_string()))
-            }
+        None => match request.headers().get("refresh-token") {
```

Comment:
Instead of a nested match none, you can use a method on option to combine the 2. I'm not sure exactly which method, but something like `or_else` would probably help.

Of course, you'd need to map them to the same type, some kind of string.
If it gets too complicated, it's fine like this.

Reply:
Sorry, I tried it for like 2 hours and I could figure it out. I'm sure that there is a way, but I'm most certain that it will be more complex that what I have now.

Reply:
I got something working, but I'm not sure if it is better than what I had before

Reply:
nvm, the str pointer were already empty when I was trying to read them

Reply:
How about:
```rust
    match (
      request.cookie("refresh_token"),
      request.headers().get("refresh-token")
    ) {
      (Some(c), _) => parse_refresh_token(c.value()),
      (_, Some(t)) => parse_refresh_token(t.to_str().unwrap()),
      (None, None) => Err(HttpResponse::Unauthorized().body("Missing refresh token")),
```

I think it captures well the logic.